internal/controller/streams.go

```diff
@@ -197,105 +164,225 @@ func (sm *streamManager) BootUpStreamAndReconnect(ctx context.Context, cancel co
+	defer logCancel()
+
+	SendLogsStream, err := sm.streamClient.client.SendLogs(logCtx)
+	if err != nil {
+		logger.Errorw("Failed to connect to server", "error", err)
+		return err
+	}
+
+	sm.streamClient.logStream = SendLogsStream
+	sm.bufferedGrpcSyncer.UpdateClient(sm.streamClient.logStream, sm.streamClient.conn)
+
+	err = sm.StreamLogs(logCtx, logCancel)
+	if err != nil {
+		logger.Errorw("Failed to bootup and stream logs", "error", err)
+		return err
+	}
+
+	return nil
+}
+
+// Generic function to manage any stream with backoff and reconnection logic.
+func manageStream(logger *zap.SugaredLogger, connectAndStream func(*zap.SugaredLogger, *streamManager) error, sm *streamManager, done chan struct{}) {
 	var backoff = 1 * time.Second
-	authn := Authenticator{Logger: logger}
 	max := big.NewInt(3)
+	rebootCounter := 0
+	// Need a way to prevent operators from all booting at the same time globally if CS has outtage. Maybe a random delay here that gives us some jitter?
 	for {
-		// Generate a random number
-		randomInt, err := rand.Int(rand.Reader, max)
+		err := connectAndStream(logger, sm)
 		if err != nil {
-			logger.Errorw("Could not generate a random int", "error", err)
-			continue
-		}
-		result := randomInt.Int64()
-		sleep := 1*time.Second + backoff + time.Duration(result)*time.Millisecond // Add randomness
-		logger.Infow("Failed to establish connection; will retry", "delay", sleep)
-		time.Sleep(sleep)
-		backoff = backoff * 2 // Exponential increase
-		clientID, clientSecret, err := authn.ReadCredentialsK8sSecrets(ctx, envMap.ClusterCreds)
-		if err != nil {
-			logger.Errorw("Could not read K8s credentials", "error", err)
-		}
-		if clientID == "" && clientSecret == "" {
-			OnboardingCredentials, err := authn.GetOnboardingCredentials(ctx, envMap.OnboardingClientId, envMap.OnboardingClientSecret)
-			if err != nil {
-				logger.Errorw("Failed to get onboarding credentials", "error", err)
-				continue
+			if rebootCounter >= 10 {
+				done <- struct{}{}
+				return
 			}
-			responseData, err := Onboard(ctx, envMap.TlsSkipVerify, envMap.OnboardingEndpoint, OnboardingCredentials, logger)
+			logger.Errorw("Failed to establish stream connection; will retry", "error", err)
+			randomInt, err := rand.Int(rand.Reader, max)
 			if err != nil {
-				logger.Errorw("Failed to register cluster", "error", err)
-				continue
-			}
-			err = authn.WriteK8sSecret(ctx, responseData, envMap.ClusterCreds)
-			time.Sleep(1 * time.Second)
-			if err != nil {
-				logger.Errorw("Failed to write secret to Kubernetes", "error", err)
-			}
-			clientID, clientSecret, err = authn.ReadCredentialsK8sSecrets(ctx, envMap.ClusterCreds)
-			if err != nil {
-				logger.Errorw("Could not read K8s credentials", "error", err)
+				logger.Errorw("Could not generate a random int", "error", err)
 				continue
 			}
+			result := randomInt.Int64()
+			sleep := 1*time.Second + backoff + time.Duration(result)*time.Millisecond // Add randomness
```

Comment:
This jitter is only up to 3 milliseconds?! That's way too small.
It should be like 10% or 20% of the delay.
```suggestion
			sleep := backoff
			jitterPct := rand.Float64() * 0.20 // [0, 20%)
			sleep = time.Duration(float64(sleep) * (1. - jitterPct))
			if sleep < 1*time.Second {
				sleep = 1 * time.Second
			}
```